json.go

```diff
@@ -19,23 +18,42 @@ func (db *DB) ExportCollection(collectionName string, exportPath string) error {
 	if !exists {
 		return ErrCollectionNotExist
 	}
-
-	result, err := db.FindAll(query.NewQuery(collectionName))
+	q := query.NewQuery(collectionName)
+	collectionCount, err := db.Count(q)
 	if err != nil {
 		return err
 	}
 
-	docs := make([]map[string]interface{}, 0)
-	for _, doc := range result {
-		docs = append(docs, doc.AsMap())
+	f, err := os.Create(exportPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	if _, err = f.WriteString("["); err != nil {
+		return err
 	}
 
-	jsonString, err := json.Marshal(docs)
+	n := 0
+	err = db.ForEach(q, func(doc *d.Document) bool {
```

Comment:
The `Count()` and the `ForEach()` statements run in separate transactions. Thus, if new documents are inserted between them, the `ForEach()` statement will iterate a number of documents `n` > `collectionCount`, which would lead to some problem with the statement:
```
if n != collectionCount {
	jsonString += ","
}
```

To prevent this issue, we can ensure the number of documents iterated by the `ForEach()` is exactly `collectionCount` using `Limit()`:
```
err = db.ForEach(q.Limit(collectionCount), func(doc *d.Document) bool {
```

Reply:
You remind me that is can also cause problems when documents are deleted between them.

Reply:
I have a suggestion for ForEach() function, let consumer()  return error instead of bool.

Reply:
I have a suggestion for ForEach() function, let consumer() return error instead of bool.

```
func (db *DB) ForEach(q *query.Query, consumer func(_ *d.Document) error) error {
	q, err := normalizeCriteria(q)
	if err != nil {
		return err
	}

	return db.IterateDocs(q, func(doc *d.Document) error {
		if err := consumer(doc); err != nil {
			return err
		}
		return nil
	})
}
```

Reply:
I have a suggestion for ForEach() function, let consumer()  return error instead of bool. If there is any failure inside this function, we can get the error info directly 

```
func (db *DB) ForEach(q *query.Query, consumer func(_ *d.Document) error) error {
	q, err := normalizeCriteria(q)
	if err != nil {
		return err
	}

	return db.IterateDocs(q, func(doc *d.Document) error {
		if err := consumer(doc); err != nil {
			return err
		}
		return nil
	})
}
```